### ISSUE TITLE
Data Transform to de-duplicate security events

### ISSUE THREAD

**Issue by mbudge** [elastic/elasticsearch#76278]
Hi, 

I started building a data transform to de-duplicate security events. After getting the process working I've not got time to finish and maintain the software. In previous security operations teams, I found this process saves analysts lots of time when threat hunting. Instead of having to sift through mountains of events to find suspicious activity, they can focus on events which haven't previous occurred on the network, or last occurred more then several months ago. For example. if a threat actor starts running powershell commands that IT don't run as part of normal business activity, this will stand out like a sore thumb in the analyst threat hunting dashboards.  

- Scroll the transforms destination index and store the first seen and count fields in a hash map, where the key is the aggregation key/document ID
- Do a composite aggregation to get the first seen timestamp, last seen timestamp and the count. Use the top hits aggregation with sorted **asc** and size 1 settings to retrieve the first event in the window. This saves the analyst time as they don't have to pivot between indexes to find the event. 
- Iterate the composite aggregation results
- Lookup each aggregation key in the hash table to retrieve the previous first seen timestamp and count, if they exist.
- Update the events first seen timestamp and add the count if they exist in the hash table.

I found running the software every 15 minutes didn’t hammer the cluster.

Also loading the first seen timestamps and counts for each document ID in the destination index into a hash table meant the composite aggregation didn’t have to re-process every event every time it runs. When running the composite aggregation against the winlogbeat and target transform index, the aggregation would re-process every event, not just events which arrived in the last 15 minutes.

Thanks

**Comment by mbudge** [comment]
Simpler way might be to allow users to run transforms which index documents using the _create operation.

https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html

**Comment by benwtrent** [comment]
Hey @mbudge so, this is like a "Event occurrence" transform. This seems like a specialization of the [`latest` function](https://www.elastic.co/blog/how-to-use-transforms-to-track-your-most-recent-customer-orders).

Basically, you want the "earliest" event plus updating counter occurrence? Or just simply the FIRST event of such a thing?

**Comment by mbudge** [comment]
Yes earliest occurrence, last occurrence and the count would be a massive set forward. The problem with the latest transform for security is the analysts still get flooded with events, and they find sifting through these on a daily basis boring and repetitive. This is because all the normal BAU IT activity is constantly happening in the background, and the analysts want to focus on new activity which is considered higher risk. For example, a new set of process.name, process.path and process.command_line which IT or current software hasn't run.

If we could also get filtering sub-aggregations we could implement more sophisticated event management use cases into elastic.

The problem at the minute is we have to build all of this outside of elastic using Go and composite aggregations.

**Comment by hendrikmuhs** [comment]
Sounds related to #71097

Transform is a generic tool to build sessions out of events. Due to the wide range of aggregations this has some drawbacks, optimizations as described here are not always possible, e.g. for cardinality. In case you only need counts and min/max values #71097 describes a more performant way. 2 challenges here: 1st implementing the incremental update and 2nd implementing detection for it. Detection means that transform should automatically switch the way it processes events given a configuration.

Regarding the current solution:

> Also loading the first seen timestamps and counts for each document ID in the destination index into a hash table meant the composite aggregation didn’t have to re-process every event every time it runs.

Transform does not "re-process every event every time", the heuristic is described in the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/transform-checkpoints.html), it only re-processes sessions for events that have changed. I guess in your case you retrieve an event for (almost) every session for every run. This might can have an adverse performance impact, because the query is potentially more expensive and you query _all_ data anyway.

One way I suggest to improve your configuration is to introduce a `date_histogram`, you could create e.g. 15 minute buckets in addition to your `terms` grouping. To get the global first/last seen and the global counts you can use aggregations again, because the transform destination is just another index. This approach uses more space, however avoids a lot of re-computation. I suggest to try this out and play with different time bucket sizes. The added `date_histogram` is less lossy, you mentioned that you want to 

> focus on new activity 

With a date histogram you can see e.g. when you suddenly get more events than usual. This is not possible if you only have global counts.

Another reason for using time buckets, IMO super important: you can age out old data (deletion or moving it to cheaper object storage on cloud). 

> If we could also get filtering sub-aggregations we could implement more sophisticated event management use cases into elastic.

Transform supports the `filter` aggregation, so I am not sure what you are missing. Maybe you can elaborate a bit. Feel free to create a new issue.
